### PR TITLE
Dashboard multi-param filters

### PR DESCRIPTION
**Pull request by joneath**
Implements issue #3736

### CODE REVIEW

**Comment by joneath** [comment]
Here's where we're at by removing the single filter param select with the new `SelectPicker` component. There's still UI work to be done to make it match the mocks (https://github.com/metabase/metabase/issues/3736)

screen shot 2017-12-12 at 14 32 16
screen shot 2017-12-12 at 14 32 29

**Comment by attekei** [comment]
My quick review while in London Gatwick airport 🇬🇧:

* If you have already applied a filter, removing the filter by unchecking the applied values and clicking "Done" leaves the filter in a weird state
screen shot 2017-12-18 at 22 42 46

* Chart drill-through doesn't work properly neither with a single selection (which is a regression) or multiple selections. For some reason, the category array is converted to a serialized string
screen shot 2017-12-18 at 22 46 08

* A strange filter value when returning to the dashboard (with browser back button) after drilling through to a question
screen shot 2017-12-18 at 22 49 11

* I didn't find any issues regarding public dashboards / embedding 🎉

* Also variable filters for native questions seem to work with multiple values. I mean this kind of scenario:

screen shot 2017-12-18 at 23 04 47

File: frontend/src/metabase/query_builder/components/filters/pickers/SelectPicker.jsx
Comment (attekei) [inline frontend/src/metabase/query_builder/components/filters/pickers/SelectPicker.jsx]: probably an accidental change?

File: frontend/test/parameters/components/widgets/CategoryWidget.integ.spec.js
Comment (attekei) [inline frontend/test/parameters/components/widgets/CategoryWidget.integ.spec.js]: Nicely written tests! 😊

This could include a test about this bug I wrote to earlier comment: "If you have already applied a filter, removing the filter by unchecking the applied values and clicking "Done" leaves the filter in a weird state"

**Review by attekei — COMMENTED** [review]
Two minor suggestions, generally LGTM

File: frontend/src/metabase/query_builder/components/filters/pickers/SelectPicker.jsx
Comment (joneath) [reply frontend/src/metabase/query_builder/components/filters/pickers/SelectPicker.jsx]: Sadly no, `CheckBox` requires a `onChange` handler to be passed. Since we're using the surrounding label as our toggle handler we don't require a `onChange` to be defined on the `CheckBox` but the `CheckBox` component does require it. I'm happy to look into removing the `onChange` handler requirement from the `CheckBox` component!

**Comment by joneath** [comment]
@attekei thanks for catching the empty values bug. I'm unable to reproduce the the question query builder being broken and the weird back button behavior. Can you provide steps to reproduce?

**Comment by salsakran** [comment]
@joneath there's a lint failure that prevents merging.

**Comment by mazameli** [comment]
Went through this pretty thoroughly just now, with a focus on seeing how this works with SQL/native cards and field filters.

- There's a phantom `Find a value` string hanging out that should be there:
screen shot 2017-12-21 at 10 43 05 am

screen shot 2017-12-21 at 11 30 51 am

- On a dashboard with some values selected in a checkbox filter, clicking through on a filtered GUI card results in an error:

screen shot 2017-12-21 at 10 44 03 am

- If a SQL/native question includes a filter widget, then clicking through to it from a filtered dashboard results in weirdness:

screen shot 2017-12-21 at 10 55 09 am

- Just to double-check, these checkboxes should be present for the category types that we treat specially as well (e.g. State, City, ID, etc.) provided that they don't exceed the number of distinct values limit that causes the widget to switch over to an input box.

**Comment by mokra** [comment]
Sorry if i didn't find the answer to my question: 
Where/How is the max. amount of data shown in the filter defined? Will it also display 1000 rows?

**Comment by mazameli** [comment]
Seems like the issues in my last comment were all addressed — nice!

The one thing I am still seeing is that clicking through on a SQL card when multiple dashboard filter selections have been made doesn't work:

screen shot 2017-12-27 at 5 19 35 pm

That above state should say `2 selections` in the filter box, and the query should return results.

**Comment by mazameli** [comment]
Happy new year, @joneath. Let me know if I can help you debug that remaining issue I mentioned above. Thanks,.

File: frontend/src/metabase/meta/types/Parameter.js
Comment (joneath) [inline frontend/src/metabase/meta/types/Parameter.js]: I had to add this extraneous type over just adding the Array type to `ParameterValue` due to flow blowing up in other parts of `frontend/src/metabase/meta/Parameter.js`. I kind of hate this so I'm very much open to suggestions.

**Comment by mazameli** [comment]
Looks like you fixed that click-through issue — nice!

I believe this is ready to merge from a UI/UX perspective.

**Comment by salsakran** [comment]
@joneath there's a merge conflict now, I think if that gets resolved we should merge this before master drifts any further.

**Comment by mazameli** [comment]
Looks like the status of this is that @joneath needs to resolve a merge conflict, and then we need a final code review+approval.

**Comment by joneath** [comment]
Conflict fixed and now ready for final review @attekei.

**Comment by joneath** [comment]
I'm now getting a failure on `Calendar.unit.test.js`. From the looks of it `Calendar.unit.test.js` is not mocking the system clock so the snapshot which has the 12th of the month as _today_ is now failing as the system clock is being used (currently the 16th of the month). I'm assuming I should fix on master and then pull down here? I'll get started on a PR now while I wait for a reply.

**Comment by joneath** [comment]
Fixed `Calendar.unit.test.js` in #6706.

File: frontend/src/metabase/lib/urls.js
Comment (tlrobinson) [inline frontend/src/metabase/lib/urls.js]: I think `kv[0]` and `v` need to be `encodeURIComponent`'d

Comment (tlrobinson) [reply frontend/src/metabase/lib/urls.js]: Also some tests with values that include special characters like `&` and `=` would be good.

File: frontend/src/metabase/query_builder/components/filters/pickers/SelectPicker.jsx
Comment (tlrobinson) [reply frontend/src/metabase/query_builder/components/filters/pickers/SelectPicker.jsx]: Looks like we actually check for `props.onClick` before calling it, so all you should need to do is remove `isRequired` from the propType.

**Review by tlrobinson — COMMENTED** [review]
Frontend looks good except for the few things I mentioned.

**Comment by joneath** [comment]
@tlrobinson I addressed your feedback in my most recent push. Thanks for the review!

**Review by tlrobinson — APPROVED** [review]

**Comment by lucasloami** [comment]
Was the docker image updated with this merge?

**Comment by salsakran** [comment]
@lucasloami we haven't released 0.28 officially yet, but the release candidate images should have this commit.

**Comment by lucasloami** [comment]
Awesome. When do you guys are planning to launch 0.28? Do you already have a schedule for this, @salsakran ?

**Comment by davehowell** [comment]
Is there a way for this Field Filter to work with URL parameters? Or some other way to use URL parameters to pass in multiple values/ an array?

I've replied on a discussion with another user asking about it and sharing my workarounds, but I feel like `Field Filter` should be a way to make it work

http://discourse.metabase.com/t/multi-select-filters-with-sql-when-embedded/3017

**Pull request merged**
